### PR TITLE
Refactor Mirror to reduce metadata allocation

### PR DESCRIPTION
**Pull request by airspeedswift**
Mirror uses `AnyCollection` to type erase the `children` between different custom mirror implementations and it's own default implementation. It was previously creating an `AnyCollection` to store the children on creation, and this was quite costly both in object and metadata creation.

This switches to an internal `Either` type to store the default children implementation, and only resorts to `AnyCollection` for custom implementations. This avoids the box allocation if you don't access the children. Unfortunately `Mirror.children: AnyCollection<Any>` is ABI, but internally the standard library can take a faster path.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Review by mikeash — APPROVED** [review]

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci smoke test linux

File: stdlib/public/core/EitherSequence.swift [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+internal typealias _EitherRandomAccessCollection<
+  L: RandomAccessCollection, R: RandomAccessCollection
+> = _Either<L,R> where L.Element == R.Element
+
+extension _EitherRandomAccessCollection: RandomAccessCollection { }
+
+extension _Either {
+  init<T, C: Collection>(
+    _ collection: C
+  ) where Right == AnyCollection<T>, C.Element == T {
```
Comment (AnthonyLatsis) [inline stdlib/public/core/EitherSequence.swift:178]: I think we can drop `T` by expressing this as `Right == AnyCollection<C.Element>`

**Pull request merged**